sync-core/src/main/java/com/cloudant/sync/datastore/encryption/EncryptedAttachmentOutputStream.java

```diff
@@ -0,0 +1,196 @@
+    protected EncryptedAttachmentOutputStream(OutputStream out) {
+        super(out);
+        cipherOutputStream = new CipherOutputStream(out, new NullCipher());
+    }
+
+    /**
+     * <p>Creates an output stream with a key and IV.</p>
+     *
+     * <p>This constructor writes bytes to the out parameter in order to write the header.</p>
+     *
+     * <p>Note: if the specified output stream is null, a NullPointerException may be thrown
+     * later when it is used.</p>
+     *
+     * @param out the output stream object.
+     * @param key the encryption key to use. Length must be supported by underlying
+     *            JCE implementation.
+     * @param iv the initialisation vector to use.
+     *
+     * @throws InvalidAlgorithmParameterException if IV is wrong size
+     * @throws InvalidKeyException if key is wrong size
+     * @throws IOException on I/O exceptions
+     */
+    public EncryptedAttachmentOutputStream(OutputStream out, byte[] key, byte[] iv)
+            throws InvalidAlgorithmParameterException, InvalidKeyException, IOException {
+        super(out);
+
+        // Don't change under our feet
+        byte[] keyCopy = Arrays.copyOf(key, key.length);
+        byte[] ivCopy = Arrays.copyOf(iv, iv.length);
+
+        try {
+
+            // Be sure Cipher is valid with passed parameters before writing anything
+            Cipher c = Cipher.getInstance(EncryptionConstants.CIPHER);
+            c.init(Cipher.ENCRYPT_MODE,
+                    new SecretKeySpec(keyCopy, EncryptionConstants.KEY_ALGORITHM),
+                    new IvParameterSpec(ivCopy));
+
+            // Write header
+            out.write(new byte[]{EncryptionConstants.ATTACHMENT_DISK_VERSION});
+            out.write(ivCopy);
+
+            // Ready to write the encrypted body
+            cipherOutputStream = new CipherOutputStream(out, c);
+
+        } catch (NoSuchPaddingException ex) {
+            // Should not happen, padding should be supported by every JCE, so wrap in RuntimeEx
+            throw new RuntimeException("Couldn't initialise crypto engine", ex);
+        } catch (NoSuchAlgorithmException ex) {
+            // Should not happen, AES should be supported by every JCE, so wrap in RuntimeException
+            throw new RuntimeException("Couldn't initialise crypto engine", ex);
+        }
+    }
+
+    @Override
+    /**
+     * <p>Closes this output stream and releases any system resources associated with this
+     * stream.</p>
+     *
+     * <p>This method invokes the doFinal method of the encapsulated cipher object, which
+     * causes any bytes buffered by the encapsulated cipher to be processed. The result is
+     * written out by calling the flush method of this output stream.</p>
+     *
+     * <p>This method resets the encapsulated cipher object to its initial state and calls
+     * the close method of the underlying output stream.</p>
+     *
+     * <p>This just proxies to the underlying CipherOutputStream.</p>
```

Comment:
I feel like this would make more sense if the comment order was reversed e.g. "This method proxies to the underlying CipherOutputStream which in turn invokes the doFinal..."
Or maybe after declaring that it proxies to CipherOutputStream just use `@see CipherOutputStream#close()`